confirmation/migrations/0008_confirmation_expiry_date.py

```diff
@@ -0,0 +1,52 @@
+# Generated by Django 3.1.7 on 2021-03-31 20:47
+
+from datetime import timedelta
+
+from django.conf import settings
+from django.db import migrations, models
+from django.db.backends.postgresql.schema import DatabaseSchemaEditor
+from django.db.migrations.state import StateApps
+
+
+def set_expiry_date_for_existing_confirmations(
+    apps: StateApps, schema_editor: DatabaseSchemaEditor
+) -> None:
+    Confirmation = apps.get_model("confirmation", "Confirmation")
+    # The values at the time of this migration
+    INVITATION = 2
+    UNSUBSCRIBE = 4
+    MULTIUSE_INVITE = 6
+
+    for confirmation in Confirmation.objects.all():
+        if confirmation.type in (INVITATION, MULTIUSE_INVITE):
+            confirmation.expiry_date = confirmation.date_sent + timedelta(
+                settings.INVITATION_LINK_VALIDITY_DAYS
+            )
+        elif confirmation.type == UNSUBSCRIBE:
+            confirmation.expiry_date = confirmation.date_sent + timedelta(1000000)
+        else:
+            confirmation.expiry_date = confirmation.date_sent + timedelta(
+                settings.CONFIRMATION_LINK_DEFAULT_VALIDITY_DAYS
+            )
+        confirmation.save()
```

Comment:
How carefully have you tested this migration?  It reads correct to me, but I'm a bit worried.

The other thing is that we have millions of these objects in production on Zulip Cloud; we will likely want to split this commit to a multi-part migration:
* First commit adds `expiry_date` as a nullable column, and adds code to write to both the old and new columns for new links.
* Next commit does this migration logic for any rows with a `null` value for `expiry_date`.  We may also want to batch the migration (doing 1K rows as a time with bulk update operations) to involve fewer database RTTs, with `atomic=False`.
* Next commit switches which data source we use to the `expiry_date` table.
* Next commit makes the `expiry_date` column no longer nullable and deletes the old table.

That way, one can deploy the first commit safely to production, and then run the migration without any downtime, do some auditing/verification, and then deploy the last two commits to finish the production deployment at scale.